### DISCUSSION TITLE
problem with multi channel audio

### DISCUSSION THREAD

**Discussion by Atefeh197** [espnet/espnet#4970] [General]
Hi

I am trying to enhance some audio files with ESPnet speech enhancement. When I try one of them with a single channel model, I get this error "AssertionError: Currently only support single channel input".

And when I try with the multi-channel model, I get this error: 

" File "/home/aaaa/.local/lib/python3.8/site-packages/espnet2/enh/layers/dnn_beamformer.py", line 243, in forward
    enhanced, ws = self.apply_beamforming(
  File "/home/aaaa/.local/lib/python3.8/site-packages/espnet2/enh/layers/dnn_beamformer.py", line 403, in apply_beamforming
    u[..., self.ref_channel].fill_(1)
IndexError: index 3 is out of bounds for dimension 1 with size 2 "

Do you know how I can handle it?

Thank you

**Comment by Emrys365** [comment]
Hi @Atefeh197, could you share the configuration file you used for training?

**Reply by Atefeh197** [reply]
Thank you, 

this is the configuration for a single channel:

cfg = d.download_and_unpack("espnet/Wangyou_Zhang_chime4_enh_train_enh_conv_tasnet_raw")
enh_model_sc = SeparateSpeech(
  train_config=cfg["train_config"],
  model_file=cfg["model_file"],
  normalize_segment_scale=False,
  show_progressbar=True,
  ref_channel=4,
  normalize_output_wav=True,
)

and this is the configuration for multi-channel:

tag = "espnet/Wangyou_Zhang_chime4_enh_train_enh_beamformer_mvdr_raw"
cfg = d.download_and_unpack(tag)
enh_model_mc = SeparateSpeech(
  train_config=cfg["train_config"],
  model_file=cfg["model_file"],
  normalize_segment_scale=False,
  show_progressbar=True,
  ref_channel=4,
  normalize_output_wav=True,
)


actually, I use a code on google colab with this link:
https://colab.research.google.com/drive/1fjRJCh96SoYLZPRxsjF9VDv4Q2VoIckI#scrollTo=RGAs52KVXWV-

**Reply by Emrys365** [reply]
Thanks! I think the problem is that your audio has <4 channels, while the downloaded model [`espnet/Wangyou_Zhang_chime4_enh_train_enh_beamformer_mvdr_raw`](https://huggingface.co/espnet/Wangyou_Zhang_chime4_enh_train_enh_beamformer_mvdr_raw/blob/main/exp/enh_train_enh_beamformer_mvdr_raw/config.yaml#L146) has the attribute `ref_channel=3`. Unfortunately, the current code in [espnet2/bin/enh_inference.py](https://github.com/espnet/espnet/blob/master/espnet2/bin/enh_inference.py#L170-L175) does not cover this case. A temporary workaround is to manually run the following lines after loading the `SeparateSpeech` model:

```python
if hasattr(enh_model.separator, "beamformer"):
    enh_model.separator.beamformer.ref_channel = ref_channel
```

<del>I will make a PR to fix this issue.</del> The PR has been merged now.

**Reply by Atefeh197** [reply]
Thank you Wangyou 

I added the below code following lines after loading:

if hasattr(enh_model.separator, "beamformer"):
    enh_model.separator.beamformer.ref_channel = ref_channel

but it still has the below error:

  File "/home/atefeh/.local/lib/python3.8/site-packages/espnet2/enh/layers/dnn_beamformer.py", line 403, in apply_beamforming
    u[..., self.ref_channel].fill_(1)
IndexError: index 2 is out of bounds for dimension 1 with size 2

Do you have any idea?

**Reply by Emrys365** [reply]
Did you try a smaller value for the argument `ref_channel` (say ref_channel=0). If it is larger than the actual number of channels in the input audio, the above error will occur.
